design/proposal-crossplane-v2.md

```diff
@@ -0,0 +1,948 @@
+different. I propose namespaced MRs have a different API group format -
+`<service>.<provider>.m.crossplane.io`. For example
+`compute.gcp.m.crossplane.io`. The `m` is short for “managed”.
+
+A refresher on how providers are implemented will help understand what changes
+provider maintainers would need to make to support namespaced MRs.
+
+Each Crossplane MR implementation consists of the following:
+
+* A Go struct representing the MR
+* The `managed.Reconciler` MR reconciler from
+  [crossplane-runtime](https://github.com/crossplane/crossplane-runtime)
+* An implementation of the MR reconciler’s `ExternalClient` interface
+
+A provider consists of a `main.go` that sets up clients and caches, and starts a
+controller using the MR reconciler and `ExternalClient` for each type of MR.
+
+Each MR Go struct is annotated with [kubebuilder annotations](https://book.kubebuilder.io/reference/markers/crd),
+which are used by controller-tools to generate CRD manifests.
+
+All MR Go structs satisfy crossplane-runtime’s
+[`resource.Managed` interface](https://pkg.go.dev/github.com/crossplane/crossplane-runtime@v1.18.0/pkg/resource#Managed).
+This interface defines getters and setter methods for features common to all
+MRs, like status conditions and management policies. MR Go structs satisfy the
+interface by:
+
+1. Embedding structs from crossplane-runtime (e.g. `xpv1.ResourceSpec`) in their
+   spec and status
+2. Generating getter and setter methods using
+   [crossplane-tools](https://github.com/crossplane/crossplane-tools)
+
+The MR reconciler works with any struct that satisfies the `resource.Managed`
+interface. It orchestrates an `ExternalClient` implementation. An
+`ExternalClient` is a set of create, read, update, and delete (CRUD) methods
+that operate on a specific type of MR struct. Given an MR struct as input, they
+either create, read, update, or delete the resource it represents in an external
+system like AWS.
+
+Upjet providers are the exception to this rule. Upjet MR structs also satisfy
+the `resource.Terraformed` interface, using generated getters and setter
+methods. Upjet providers use a single `ExternalClient` implementation that works
+with any MR Go struct that satisfies this interface.
+
+Given this context, the process to build a namespaced provider would be:
+
+1. Duplicate all MR structs
+    1. Remove the `// +kubebuilder:resource:scope=Cluster` comment marker
+    2. Change the API group constant to end in `m.crossplane.io`
+    3. Change to a namespaced variant of the embedded `xpv1.ResourceSpec`
+    4. Re-run crossplane-tools and Upjet code generation against the updated
+       types
+2. Duplicate all `ExternalClient` implementations
+    5. Change the duplicated implementations to use the duplicated structs
+3. Duplicate main.go and a few “plumbing” files that register types and
+   construct controllers
+
+Upjet providers can skip step 2, as there are no unique `ExternalClient`
+implementations to duplicate.
```

Comment:
I'm planning to brainstorm with some of the Upjet maintainers to see if we can find a more automated approach. Ideally provider maintainers would need to do little to nothing to support namespaced MRs.

Reply:
I've opened https://github.com/crossplane/crossplane/issues/6280 to track the problem of introducing namespaced providers, and some of the options.